examples/function_router/src/app.rs

```diff
@@ -0,0 +1,118 @@
+use yew::prelude::*;
+use yew_router::prelude::*;
+
+use crate::components::nav::Nav;
+use crate::pages::{
+    author::Author, author_list::AuthorList, home::Home, page_not_found::PageNotFound, post::Post,
+    post_list::PostList,
+};
+
+#[derive(Routable, PartialEq, Clone, Debug)]
+pub enum Route {
+    #[at("/posts/:id")]
+    Post { id: u32 },
+    #[at("/posts")]
+    Posts,
+    #[at("/authors/:id")]
+    Author { id: u32 },
+    #[at("/authors")]
+    Authors,
+    #[at("/")]
+    Home,
+    #[not_found]
+    #[at("/404")]
+    NotFound,
+}
+
+#[function_component]
+pub fn App() -> Html {
+    html! {
+        <BrowserRouter>
+            <Nav />
+
+            <main>
+                <Switch<Route> render={Switch::render(switch)} />
+            </main>
+            <footer class="footer">
+                <div class="content has-text-centered">
+                    { "Powered by " }
+                    <a href="https://yew.rs">{ "Yew" }</a>
+                    { " using " }
+                    <a href="https://bulma.io">{ "Bulma" }</a>
+                    { " and images from " }
+                    <a href="https://unsplash.com">{ "Unsplash" }</a>
+                </div>
+            </footer>
+        </BrowserRouter>
+    }
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+mod arch_native {
+    use super::*;
+
+    use yew::virtual_dom::AttrValue;
+    use yew_router::history::{AnyHistory, History, MemoryHistory};
+
+    use std::collections::HashMap;
+
+    #[derive(Properties, PartialEq, Debug)]
+    pub struct ServerAppProps {
+        pub url: AttrValue,
+        pub queries: HashMap<String, String>,
+    }
+
+    #[function_component]
+    pub fn ServerApp(props: &ServerAppProps) -> Html {
+        let history = AnyHistory::from(MemoryHistory::new());
+        history
+            .push_with_query(&*props.url, &props.queries)
+            .unwrap();
+
+        html! {
+            <Router history={history}>
+                <Nav />
+
+                <main>
+                    <Switch<Route> render={Switch::render(switch)} />
+                </main>
+                <footer class="footer">
+                    <div class="content has-text-centered">
+                        { "Powered by " }
+                        <a href="https://yew.rs">{ "Yew" }</a>
+                        { " using " }
+                        <a href="https://bulma.io">{ "Bulma" }</a>
+                        { " and images from " }
+                        <a href="https://unsplash.com">{ "Unsplash" }</a>
+                    </div>
+                </footer>
+            </Router>
+        }
+    }
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+pub use arch_native::*;
+
+fn switch(routes: &Route) -> Html {
+    match routes.clone() {
+        Route::Post { id } => {
+            html! { <Post seed={id} /> }
+        }
```

Comment:
How would we generate the content on server and pass it in instead of passing in a seed and generating the content on the client? I'm thinking of functionality similar to [`getServerSideProps` from NextJS](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props)

Reply:
I am not familiar with NextJS but the current way of data fetching is `<Suspense />` (fetch twice on both client and server).

Unless we make this a macro-based hook (which is not possible), it will be a burden on the user to gate it with feature flags for proper tree shaking. I think unless you are fetching data maybe simply make the logic to create the same state on the client side is easier?

Maybe this can be a part of a future PR alongside macro-based hooks and `useId`?

Reply:
What if I want to only fetch the data once? It doesn't make sense to load the data twice, if the client doesn't need to listen to changes.

Think of a file being displayed. There's no need to load the file again on the client. If the file is changed, the page is reloaded to view the new contents

Reply:
> What if I want to only fetch the data once?

Unfortunately, this PR does not include such a utility at this moment.
Can we make this a part of a future PR?

There're some bits missing that prevents implementing a hook that passes information from server to client easily in this PR (I think this pull request is big enough):

1. ability to recover information from comment nodes.
2. macro-based hooks (so that server side logics can be tree shaked.)

> There's no need to load the file again on the client. 

There are cases where this makes sense.
However, prevent refetching the file separately does not always lead to a better user experience especially when the file is big.
(embedding it in the same page prolongs the download time of the html file and hence prevents hydration of all components, not just that 1 component.)

Reply:
That is true, I would still like to have the option of fetching data on the server and displaying it. Even if it's just fetching on the server and passing it as a prop to the page.

Here's an example usage of this:
Data is fetched in the component and passed as prop: https://github.com/hamza1311/dear-diary-web/blob/5f28fe308db3a7c1d9fdac425ffdd1457a65789a/src/pages/%5Bid%5D.tsx#L145-L161  
Data is received as prop: https://github.com/hamza1311/dear-diary-web/blob/5f28fe308db3a7c1d9fdac425ffdd1457a65789a/src/pages/%5Bid%5D.tsx#L69-L70

In my example, the object is wrapped in an intermediary type so the client can easily interface with firebase SDK.

When the prop is passed, page is rendered initially on the server side and then HTML is sent, alongside JS which is used to hydrate the page. You can see an example page on the site at https://dear-diary.mhamza.dev/TNJlenEPG8xIwmtFFoeO

This is not supposed to be usable by public so there's no sign up page but if you want to sign in and see how it's used, let me know.

Reply:
I should mention: it's something that should be implemented but doesn't _have to be_ in this PR

Reply:
What's the advantage of SSR a protected page?

Reply:
One advantage I can think of is load times. When the authentication is done by the server and result is sent back, the client doesn't have to go through multiple steps to handle it. That also reduces the code complexity on client side and there's no need to display a loading indicator. 

In my case, I did it because why not and I wanted to try out next js

Reply:
Hmm, I see. Thank you.